File: snap/lib/graphics/CExtensions/SnapMetrics.c

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <float.h>

#define snap_fuzzy_is_null(NUM) ((NUM >= -SNAP_DOUBLE_MIN && NUM <= SNAP_DOUBLE_MIN) ? 1:0)

#define SNAP_DOUBLE_MIN DBL_MIN //0.000000000001
#define SNAP_DOUBLE_MAX DBL_MAX

#define SNAP_MIN(A, B) ((A < B) ? A : B)
#define SNAP_MAX(A, B) ((A > B) ? A : B)
#define SNAP_ABS(NUM) ((NUM < 0) ? NUM * -1 : NUM)

// XXX TODO this is duplicated from the SnapMatrix support lib
double SNAP_IDENTITY_MATRIX[16] = {
	1.0, 0.0, 0.0, 0.0,
	0.0, 1.0, 0.0, 0.0,
	0.0, 0.0, 1.0, 0.0,
	0.0, 0.0, 0.0, 1.0
};

void snap_matrix_transform_point(double* matrix, double* point, double delta, double* ANSWER){

	// delta is int 0 or 1 but making it double so it mixes into calculations
	// NOTE: this means set delta to 1.0 to include translation and to 0.0 to exclude translation

	// https://www.scratchapixel.com/lessons/mathematics-physics-for-computer-graphics/geometry/transforming-points-and-vectors

	/*
	point represents the translation component of another matrix that we want to map into or multiply by this matrix

	1 0 0 x
	0 1 0 y
	0 0 1 z
	0 0 0 1

	00 01 02 03
	10 11 12 13
	20 21 22 23
	30 31 32 33
	*/

	double m00 = matrix[0];
	double m01 = matrix[1];
	double m02 = matrix[2];
	double m03 = matrix[3];

	double m10 = matrix[4];
	double m11 = matrix[5];
	double m12 = matrix[6];
	double m13 = matrix[7];

	double m20 = matrix[8];
	double m21 = matrix[9];
	double m22 = matrix[10];
	double m23 = matrix[11];

	//double m30 = matrix[12];
	//double m31 = matrix[13];
	//double m32 = matrix[14];
	//double m33 = matrix[15];

	// unpack vector
	double v0 = point[0];
	double v1 = point[1];
	double v2 = point[2];
	//double v3 = 1;//vec3[3];

	// position is column vector (translation), need to multiply by rows of matrix!
	ANSWER[0] = m00*v0 + m01*v1 + m02*v2 + m03*delta;//*v3;
	ANSWER[1] = m10*v0 + m11*v1 + m12*v2 + m13*delta;//*v3;
	ANSWER[2] = m20*v0 + m21*v1 + m22*v2 + m23*delta;//*v3;
	//ANSWER[3] = m30*v0 + m31*v1 + m32*v2 + m33*delta;//*v3;

	// was ANSWER[idx]: 3, 7, 11, 15
	//ANSWER[0] = m03*b00 + m13*b01 + m23*b02 + m33*b03;
	//ANSWER[1] = m03*b10 + m13*b11 + m23*b12 + m33*b13;
	//ANSWER[2] = m03*b20 + m13*b21 + m23*b22 + m33*b23;
	//ANSWER[3] = m03*b30 + m13*b31 + m23*b32 + m33*b33;

	//ANSWER[0] = m03*1 + m13*0 + m23*0 + m33*v0;
	//ANSWER[1] = m03*0 + m13*1 + m23*0 + m33*v1;
	//ANSWER[2] = m03*0 + m13*0 + m23*1 + m33*v2;
	//ANSWER[3] = m03*0 + m13*0 + m23*0 + m33*v3;

	//ANSWER[0] = m03 + m33*v0;
	//ANSWER[1] = m13 + m33*v1;
	//ANSWER[2] = m23 + m33*v2;
	//ANSWER[3] = m33*v3;
}




#define _snap_extents_are_null(EXTENTS) (\
	(EXTENTS != NULL && ( \
	/*instead of fuzzy_is_null, since value must be positive (min <= max must be true)*/\
	EXTENTS[3]-EXTENTS[0] > SNAP_DOUBLE_MIN || \
	EXTENTS[4]-EXTENTS[1] > SNAP_DOUBLE_MIN || \
	EXTENTS[5]-EXTENTS[2] > SNAP_DOUBLE_MIN )  \
	) ? 0 : 1)/*note truth switch, since this is a negative statement*/

int snap_extents_are_null(double* EXTENTS){
	return _snap_extents_are_null(EXTENTS);
}

void snap_matrix_map_extents(double *matrix, double *extents, int is_delta, double *ANSWER){
	// NOTE: use inverted matrix to map back out

	snap_matrix_transform_point(matrix, extents, (double)!is_delta, ANSWER);
	snap_matrix_transform_point(matrix, extents + 3, (double)!is_delta, ANSWER + 3);

	// rotations may re-assign extents min/max order, make sure min <= max
	double tmp;
	int idx = -1;
	while (++idx < 3){

		if (ANSWER[idx] <= ANSWER[idx+3])
			continue;

		tmp = ANSWER[idx];
		ANSWER[idx] = ANSWER[idx+3];
		ANSWER[idx+3] = tmp;
	}
}

#define _snap_extents_contain_point(EXTENTS, POINT) (\
	(EXTENTS != NULL && POINT != NULL && (\
	/* extents are always aligned to axes, they have no concept of rotation
	also it is assumed max point is >= min point ALWAYS, it is an error otherwise
	but will just evaluate to outside of rect if min >= max*/\
	POINT[0] >= EXTENTS[0] && POINT[0] <= EXTENTS[3] && \
	POINT[1] >= EXTENTS[1] && POINT[1] <= EXTENTS[4] && \
	POINT[2] >= EXTENTS[2] && POINT[2] <= EXTENTS[5]) \
	) ? 1 : 0)

int snap_extents_contain_point(double* EXTENTS, double* POINT){
	return _snap_extents_contain_point(EXTENTS, POINT);
}



int snap_extents_diffmatrix(double* BASE_ext, double* ITEM_ext, int lock_scale, int uniform, double align_x, double align_y, double align_z, double* MATRIX_ANSWER){//, SnapNode* MSG){

	// calculates the matrix that would transform ITEM_ext to fit into BASE_ext
	// NOTE: the extents should have been mapped from their container matrices, and are being compared as peers

	if (!MATRIX_ANSWER){
		//snap_error("%s no matrix to write answer into!", SNAP_FUNCTION_NAME);
		return 1;//(any)"ERROR";
	}
	if (!(BASE_ext && ITEM_ext)){
		memcpy(MATRIX_ANSWER, SNAP_IDENTITY_MATRIX, 16 * sizeof (double));
		return 0;
	}

	//int uniform = 1;
	// centered (this will only have an effect if the object axis does not fill the allocation)
	//double align[] = {.5, .5, .5}; // TODO handle oversize properly too?
	double align[] = {align_x, align_y, align_z};
	// TODO align is a weighting, if 1 it means align the right side with the right side, if 0 it means align the left with the left, if .5 it means centered (regardless of scale, if bigger or smaller the side alignment is the same result...)

	/*
	{for_attr_in_SnapNode(MSG)
		if (attr == (any)"uniform"){
			uniform = snap_bool_as_int(value);
		}
		else if (attr == (any)"align"){
			snap_memcpy(align, value, 3 * sizeof (double));
		}
		else if (attr == (any)"x_align"){
			align[0] = *((double*)value);
		}
		else if (attr == (any)"y_align"){
			align[1] = *((double*)value);
		}
		else if (attr == (any)"z_align"){
			align[2] = *((double*)value);
		}
	}}
	*/

	double base_dims[3] = {
		BASE_ext[3] - BASE_ext[0],
		BASE_ext[4] - BASE_ext[1],
		BASE_ext[5] - BASE_ext[2],
		};


	//double item_origin[3] = {ITEM_ext[0], ITEM_ext[1], ITEM_ext[3]};
	//double *item_origin = ITEM_ext;
	double item_dims[3] = {
		SNAP_MAX(ITEM_ext[3] - ITEM_ext[0], SNAP_DOUBLE_MIN),
		SNAP_MAX(ITEM_ext[4] - ITEM_ext[1], SNAP_DOUBLE_MIN),
		SNAP_MAX(ITEM_ext[5] - ITEM_ext[2], SNAP_DOUBLE_MIN),
		};

	//snap_out("ITEM_dims w(%.2lf) h(%.2lf) d(%.2lf)", item_dims[0], item_dims[1], item_dims[2]);

	// TODO negative scaling?
	double scale[3] = {1,1,1};
	if (!lock_scale){
		// we only want to consider valid entries and there might not be any!  so flag if they are valid
		unsigned int valid_indexes = 0;
		int i = -1;
		while (++i < 3){
			//if (item_dims[i] != 0){ // TODO fuzzy is null?
			//if (item_dims[i] > 0.001){//
			if (!snap_fuzzy_is_null(item_dims[i])){
				scale[i] = SNAP_MAX(base_dims[i] / item_dims[i], SNAP_DOUBLE_MIN);
				//snap_out("scale[%d] %lf base(%lf) item(%lf)", i, scale[i], base_dims[i], item_dims[i]);
				//printf("scale[%d] %lf base(%lf) item(%lf)\n", i, scale[i], base_dims[i], item_dims[i]);
				valid_indexes |= (1 << i);
			}
		}
		if (uniform){
			// if uniform we scale uniformly by the smallest scale
			if (valid_indexes){

				// a uniform scale can only be as large as the smallest axis
				// (the one that scales the least) TODO if it gets smaller then isn't it the biggest?
				double smallest;

				switch (valid_indexes){
					// number represents binary flags in reverse
					case 1: // 001 -> 100
						smallest = scale[0];
						break;
					case 2: // 010 -> 010
						smallest = scale[1];
						break;
					case 3: // 011 -> 110
						smallest = SNAP_MIN(scale[0], scale[1]);
						break;
					case 4: // 100 -> 001
						smallest = scale[2];
						break;
					case 5: // 101 -> 101
						smallest = SNAP_MIN(scale[0], scale[2]);
						break;
					case 6: // 110 -> 011
						smallest = SNAP_MIN(scale[1], scale[2]);
						break;
					case 7: // 111 -> 111
						smallest = SNAP_MIN(scale[0], SNAP_MIN(scale[1], scale[2]));
						break;
					default:
						//snap_error("INVALID INDEXES?? %d", valid_indexes);
						return 1;//(any)"ERROR";
				}

				printf("smallest %lf\n", smallest);
				scale[0] = smallest;
				scale[1] = smallest;
				scale[2] = smallest;

			}

			// else scale is all 1's, which it should already be
		}
	}

	/*
	double tm[16] = {
		scale[0],0,0, BASE_ext[0] + ((base_dims[0] - (item_dims[0] * scale[0])) * align[0]) - (ITEM_ext[0] * scale[0]),
		0,scale[1],0, BASE_ext[1] + ((base_dims[1] - (item_dims[1] * scale[1])) * align[1]) - (ITEM_ext[1] * scale[1]),
		0,0,scale[2], BASE_ext[2] + ((base_dims[2] - (item_dims[2] * scale[2])) * align[2]) - (ITEM_ext[2] * scale[2]),
		0,0,0, 1
		};
	*/
	// 'unused'
	MATRIX_ANSWER[1] = 0.;
	MATRIX_ANSWER[2] = 0.;
	MATRIX_ANSWER[4] = 0.;
	MATRIX_ANSWER[6] = 0.;
	MATRIX_ANSWER[8] = 0.;
	MATRIX_ANSWER[9] = 0.;
	MATRIX_ANSWER[12] = 0.;
	MATRIX_ANSWER[13] = 0.;
	MATRIX_ANSWER[14] = 0.;
	MATRIX_ANSWER[15] = 1.;	

	// translation
	MATRIX_ANSWER[3] = BASE_ext[0] + ((base_dims[0] - (item_dims[0] * scale[0])) * align[0]) - (ITEM_ext[0] * scale[0]);
	MATRIX_ANSWER[7] = BASE_ext[1] + ((base_dims[1] - (item_dims[1] * scale[1])) * align[1]) - (ITEM_ext[1] * scale[1]);
	MATRIX_ANSWER[11] = BASE_ext[2] + ((base_dims[2] - (item_dims[2] * scale[2])) * align[2]) - (ITEM_ext[2] * scale[2]);

	// scale
	MATRIX_ANSWER[0] = scale[0];
	MATRIX_ANSWER[5] = scale[1];
	MATRIX_ANSWER[10] = scale[2];

	//printf("matrix(\n");
	//printf("\t%lf, %lf, %lf, %lf\n", MATRIX_ANSWER[0], MATRIX_ANSWER[1], MATRIX_ANSWER[2], MATRIX_ANSWER[3]);
	//printf("\t%lf, %lf, %lf, %lf\n", MATRIX_ANSWER[4], MATRIX_ANSWER[5], MATRIX_ANSWER[6], MATRIX_ANSWER[7]);
	//printf("\t%lf, %lf, %lf, %lf\n", MATRIX_ANSWER[8], MATRIX_ANSWER[9], MATRIX_ANSWER[10], MATRIX_ANSWER[11]);
	//printf("\t%lf, %lf, %lf, %lf\n", MATRIX_ANSWER[12], MATRIX_ANSWER[13], MATRIX_ANSWER[14], MATRIX_ANSWER[15]);
	//printf(")\n");

	return 0;
}

/* TODO put this in python, it requires class awareness...
int snap_extents_fit(double* BASE_ext, SnapNode ITEM, SnapNode* MSG){
	// fits ITEM inside of BASE_ext by translating and scaling ITEM's matrix by the difference

	double ITEM_ext[] = {0.,0.,0., 0.,0.,0.};
	snap_event(&ITEM, "GET", "extents", ITEM_ext);
	
	// XXX TODO use snap_event(self, "GET", "extents", ext);
	//double* BASE_extents_src = (double*)snap_getattr_at(self, "_extents_", IDX_SnapMetrics__extents_);
	//double* ITEM_extents_src = (double*)snap_getattr_at(&ITEM, "_extents_", IDX_SnapMetrics__extents_);

	if (snap_extents_are_null(BASE_ext) || snap_extents_are_null(ITEM_ext)){
		snap_warning("invalid extents for fit operation");
		snap_extents_print(BASE_ext);
		snap_extents_print(ITEM_ext);
		return (any)"ERROR";
	}

	// localized for assign below
	double* ITEM_matrix = (double*)snap_getattr_at(&ITEM, "_matrix_", IDX_SnapMatrix__matrix_);
	if (!ITEM_matrix){
		snap_error("no matrix on item!");
		return (any)"ERROR";
	}

	// extents are local to item, we want them in item parent space (the two are compared as peers)
	snap_map_extents(ITEM_matrix, ITEM_ext, 0, ITEM_ext);

	double tm[16];
	if (_snap_extents_diffmatrix(BASE_ext, ITEM_ext, tm, MSG) == (any)"ERROR"){
		snap_error("%s unable to calculate diffmatrix", SNAP_FUNCTION_NAME);
		return (any)"ERROR";
	}

	snap_matrix_transform(ITEM_matrix, tm, NULL, ITEM_matrix);

	return NULL;
}
*/

void snap_extents_mix(double* A, double* B, double* ANSWER){
	// combines two extents together into one
	ANSWER[0] = SNAP_MIN(A[0], B[0]);
	ANSWER[1] = SNAP_MIN(A[1], B[1]);
	ANSWER[2] = SNAP_MIN(A[2], B[2]);
	ANSWER[3] = SNAP_MAX(A[3], B[3]);
	ANSWER[4] = SNAP_MAX(A[4], B[4]);
	ANSWER[5] = SNAP_MAX(A[5], B[5]);
}

void snap_extents_ensure_correct(double* EXTENTS){
	// makes sure min < max or swaps them
	double tmp;
	if (EXTENTS[3] < EXTENTS[0]){
		tmp = EXTENTS[3];
		EXTENTS[3] = EXTENTS[0];
		EXTENTS[0] = tmp;
	}
	if (EXTENTS[4] < EXTENTS[1]){
		tmp = EXTENTS[4];
		EXTENTS[4] = EXTENTS[1];
		EXTENTS[1] = tmp;
	}
	if (EXTENTS[5] < EXTENTS[2]){
		tmp = EXTENTS[5];
		EXTENTS[5] = EXTENTS[2];
		EXTENTS[2] = tmp;
	}
}

File: snap/lib/graphics/SnapMetrics.py

def build(ENV):

	#recompile_support_library = ENV.recompile_support_library

	snap_matrix_transform = ENV.snap_matrix_transform

	SnapMatrix = ENV.SnapMatrix

	ctypes = ENV.extern.ctypes

	#snap_send = ENV.snap_send
	#snap_emit = ENV.snap_emit

	snap_warning = ENV.snap_warning

	#if ENV.__SNAP_IS_PYTHON__:

	SnapMetricsExt = ctypes.snap_recompile_support_library(__file__)

	# int snap_extents_are_null(double* EXTENTS){
	snap_extents_are_null = SnapMetricsExt.snap_extents_are_null
	snap_extents_are_null.argtypes = [ctypes.POINTER(ctypes.c_double)]
	snap_extents_are_null.restype = ctypes.c_int
	ENV.snap_extents_are_null = snap_extents_are_null

	# void snap_matrix_map_extents(double *matrix, double *extents, int is_delta, double *ANSWER){
	snap_matrix_map_extents = SnapMetricsExt.snap_matrix_map_extents
	snap_matrix_map_extents.argtypes = [ctypes.POINTER(ctypes.c_double), ctypes.POINTER(ctypes.c_double), ctypes.c_int, ctypes.POINTER(ctypes.c_double)]
	snap_matrix_map_extents.restype = None
	ENV.snap_matrix_map_extents = snap_matrix_map_extents

	# int snap_extents_contain_point(double* EXTENTS, double* POINT){
	snap_extents_contain_point = SnapMetricsExt.snap_extents_contain_point
	snap_extents_contain_point.argtypes = [ctypes.POINTER(ctypes.c_double), ctypes.POINTER(ctypes.c_double)]
	snap_extents_contain_point.restype = ctypes.c_int
	ENV.snap_extents_contain_point = snap_extents_contain_point

	# int snap_extents_diffmatrix(double* BASE_ext, double* ITEM_ext, int uniform, double align_x, double align_y, double align_z, double* MATRIX_ANSWER){
	snap_extents_diffmatrix = SnapMetricsExt.snap_extents_diffmatrix
	snap_extents_diffmatrix.argtypes = [ctypes.POINTER(ctypes.c_double), ctypes.POINTER(ctypes.c_double), ctypes.c_int, ctypes.c_int, ctypes.c_double, ctypes.c_double, ctypes.c_double, ctypes.POINTER(ctypes.c_double)]
	snap_extents_diffmatrix.restype = ctypes.c_int
	ENV.snap_extents_diffmatrix = snap_extents_diffmatrix

	# void snap_extents_mix(double* A, double* B, double* ANSWER){
	snap_extents_mix = SnapMetricsExt.snap_extents_mix
	snap_extents_mix.argtypes = [ctypes.POINTER(ctypes.c_double), ctypes.POINTER(ctypes.c_double), ctypes.POINTER(ctypes.c_double)]
	snap_extents_mix.restype = None
	ENV.snap_extents_mix = snap_extents_mix

	# void snap_extents_ensure_correct(double* EXTENTS){
	snap_extents_ensure_correct = SnapMetricsExt.snap_extents_ensure_correct
	snap_extents_ensure_correct.argtypes = [ctypes.POINTER(ctypes.c_double)]
	snap_extents_ensure_correct.restype = None
	ENV.snap_extents_ensure_correct = snap_extents_ensure_correct

	snap_extents_t = (ctypes.c_double * 6)
	ENV.snap_extents_t = snap_extents_t

	snap_matrix_map_extents = ENV.snap_matrix_map_extents
	snap_matrix_t = ENV.snap_matrix_t
	snap_matrix_invert = ENV.snap_matrix_invert


	# TODO make child_fit and parent_fit methods?  along with fit()?
	def snap_extents_fit(BASE_ext, ITEM, **SETTINGS):
		# fits ITEM inside of BASE_ext by translating and scaling ITEM's matrix by the difference

		ITEM_ext = ITEM['extents']

		# TODO make scaling optional!  
		
		# XXX TODO use snap_event(self, "GET", "extents", ext);
		#double* BASE_extents_src = (double*)snap_getattr_at(self, "_extents_", IDX_SnapMetrics__extents_);
		#double* ITEM_extents_src = (double*)snap_getattr_at(&ITEM, "_extents_", IDX_SnapMetrics__extents_);

		assert not (snap_extents_are_null(BASE_ext) or snap_extents_are_null(ITEM_ext)), "invalid extents for fit operation"

		# localized for assign below
		ITEM_matrix = ITEM['matrix']
		assert ITEM_matrix is not None, 'no item matrix?'

		# extents are local to item, we want them in item parent space (the two are compared as peers)
		snap_matrix_map_extents(ITEM_matrix, ITEM_ext, 0, ITEM_ext)

		tm = snap_matrix_t()
		align_x = SETTINGS.get('align_x', 0.5)
		align_y = SETTINGS.get('align_y')
		if align_y is None:
			align_y = align_x
		align_z = SETTINGS.get('align_z')
		if align_z is None:
			align_z = align_x
		lock_scale = SETTINGS.get('lock_scale', False)
		uniform = SETTINGS.get('uniform', True)
		if snap_extents_diffmatrix(BASE_ext, ITEM_ext, lock_scale, uniform, align_x, align_y, align_z, tm) != 0:
			raise ValueError("snap_extents_fit() unable to calculate diffmatrix")

		snap_matrix_transform(ITEM_matrix, tm, None, ITEM_matrix)

		return None

	ENV.snap_extents_fit = snap_extents_fit

	# TODO merge_extents() -> into one area?  (basically just take the biggest and smallest point)

	class SnapMetrics(SnapMatrix):

		__slots__ = []
		#__slots__ = ['_extents_']

		@ENV.SnapProperty
		class extents:

			def get(self, MSG):
				"""()->snap_extents_t"""
				e = self.__snap_data__['extents']
				if e is None:
					e = snap_extents_t(0,0,0, 0,0,0)
				#	e = self.__snap_data__['extents'] = snap_extents_t(0,0,0, 1,1,1)
				#return e
				return e

			def set(self, MSG):
				"""(snap_extents_t!)""" # TODO list|tuple or *values...
				ext = MSG.args[0]
				if ext is not None:
					assert isinstance(ext, snap_extents_t)
					if ext[3] <= ext[0]: ext[3] = ext[0]
					if ext[4] <= ext[1]: ext[4] = ext[1]
					if ext[5] <= ext[2]: ext[5] = ext[2]
					self.__snap_data__['extents'] = ext
				else:
					del self.__snap_data__['extents']
				self.changed(extents=ext)

		@ENV.SnapProperty
		class mapped_extents:
			# the reverse of this is to multiply extents by inverted self['matrix']

			def get(self, MSG):
				"""()->snap_extents_t"""
				ext = snap_extents_t(*self['extents'])
				snap_matrix_map_extents(self['matrix'], ext, 0, ext)
				return ext

		@ENV.SnapProperty
		class width:

			def get(self, MSG):
				"""()->float"""
				e = self['extents']
				return float(e[3]-e[0])

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				e = self['extents']
				if value is None:
					value = 0
				e[3] = e[0] + value
				self['extents'] = e
				self.changed(width=value)
				

		@ENV.SnapProperty
		class height:

			def get(self, MSG):
				"""()->float"""
				e = self['extents']
				return float(e[4]-e[1])

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				e = self['extents']
				if value is None:
					value = 0
				e[4] = e[1] + value
				self['extents'] = e
				self.changed(height=value)

		@ENV.SnapProperty
		class depth:

			def get(self, MSG):
				"""()->float"""
				e = self['extents']
				return float(e[5]-e[2])

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				e = self['extents']
				if value is None:
					value = 0
				e[5] = e[2] + value
				self['extents'] = e
				self.changed(depth=value)

		@ENV.SnapProperty
		class size:

			def get(self, MSG):
				"""()->tuple(float,float)"""
				e = self['extents']
				return e[3]-e[0], e[4]-e[1] # w,h

			def set(self, MSG):
				"""(
				(int|float!, int|float!),
				(tuple(int|float, int|float)!),
				(width=int|float?, height=int|float?),
				)"""
				if MSG.args:
					if len(MSG.args) == 2:
						w,h = MSG.args
					elif len(MSG.args) == 1:
						w,h = MSG.args[0]
					else:
						raise IOError('invalid args', MSG.args)
				elif MSG.kwargs:
					w,h = MSG.kwargs.get('width', self['width']), MSG.kwargs.get('height',self['height'])
				else:
					raise IOError('invalid arguments', MSG)
				e = self['extents']
				e[3] = e[0] + float(w)
				e[4] = e[1] + float(h)
				self['extents'] = e
				self.changed(size=(w,h))

		@ENV.SnapProperty
		class dimensions:

			def get(self, MSG):
				"""()->tuple(float,float,float)"""
				e = self['extents']
				return e[3]-e[0], e[4]-e[1], e[5]-e[2] # w,h,d

			def set(self, MSG):
				"""(
				(int|float!, int|float!, int|float!),
				(tuple(3 * int|float)!),
				(width=int|float?, height=int|float?, depth=int|float?),
				)"""
				if MSG.args:
					w,h,d = MSG.args[:3]
				elif MSG.kwargs:
					w,h,d = MSG.kwargs.get('width', self['width']), MSG.kwargs.get('height', self['height']), MSG.kwargs.get('depth', self['depth'])
				else:
					raise IOError('invalid arguments', MSG)
				e = self['extents']
				e[3] = e[0] + float(w)
				e[4] = e[1] + float(h)
				e[5] = e[3] + float(d)
				self['extents'] = e
				self.changed(dimensions=(w,h,d))
				

		@ENV.SnapProperty
		class bounds:
			def get(self, MSG):
				"""()->tuple(4 * float)"""
				e = self.extents()
				return e[0], e[1], e[3]-e[0], e[4]-e[1] # x,y,w,h

		# XXX TODO x,y,z should refer to matrix origin?
		@ENV.SnapProperty
		class x:
			def get(self, MSG):
				"""()->float"""
				return self['extents'][0]

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				ext = self['extents']
				if value:
					ext[0] = value
				else:
					ext[0] = 0
				self['extents'] = ext # which will clamp and emit
				self.changed(x=value)

		@ENV.SnapProperty
		class y:
			def get(self, MSG):
				"""()->float"""
				return self['extents'][1]

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				ext = self['extents']
				if value:
					ext[1] = value
				else:
					ext[1] = 0
				self['extents'] = ext # which will clamp and emit
				self.changed(x=value)

		@ENV.SnapProperty
		class z:
			def get(self, MSG):
				"""()->float"""
				return self['extents'][2]

			def set(self, MSG):
				"""(int|float!)"""
				value = MSG.args[0]
				ext = self['extents']
				if value:
					ext[2] = value
				else:
					ext[2] = 0
				self['extents'] = ext # which will clamp and emit
				self.changed(x=value)

		#def origin(self, MSG): # XXX this is ambiguous? use min_point() max_point()?  origin() is really center()?
		#	return self.extents()[:3]

		@ENV.SnapProperty
		class min_point:
			def get(self, MSG):
				"""()->tuple(3 * float)"""
				return self['extents'][:3]

		@ENV.SnapProperty
		class max_point:
			def get(self, MSG):
				"""()->tuple(3 * float)"""
				return self['extents'][3:]

		@ENV.SnapProperty
		class radius:

			def get(self, MSG):
				"""()->float"""
				# https://www.engineeringtoolbox.com/distance-relationship-between-two-points-d_1854.html
				# distance from center to largest extents (considered as a circular bounds)
				# TODO consider the angle?  technically it should be ray from center to furthest corner...

				# extents is min and max point, so just calculate .5 * vector/ray between them!
				e = self['extents']
				return float( (((e[3]-e[0])**2 + (e[4]-e[1])**2 + (e[5]-e[2])**2)**.5) * .5 )

		@ENV.SnapProperty
		class diameter:

			def get(self, MSG):
				"""()->float"""
				e = self['extents']
				return float( ((e[3]-e[0])**2 + (e[4]-e[1])**2 + (e[5]-e[2])**2)**.5 )

		@ENV.SnapProperty
		class center:

			def get(self, MSG):
				"""()->tuple(3 * float)""" # TODO use snap_vector_t?
				#SnapMetrics_proportion(self, .5, .5, .5, (double*)value);
				e = self['extents']
				return e[0] + (e[3]-e[0]) * .5, e[1] + (e[4]-e[1]) * .5, e[2] + (e[5]-e[2]) * .5

		# TODO 'left', 'right', 'top', 'bottom', also compass (north, west, east, south, ...)?


		@ENV.SnapChannel
		def fit(self, MSG):
			"""(SnapMetrics!)->snap_extents_t"""
			return snap_extents_fit(self['extents'], MSG.args[0], **MSG.kwargs)

		#def crop(self, MSG):
		#	ENV.snap_warning("crop() deprecated; use allocate()")
		#	return self.allocate.__direct__(MSG)

		@ENV.SnapChannel
		def allocateXXX(self, MSG): # XXX deprecated; use extents.set()!
			"""(extents=snap_extents_t)"""
			extents = MSG.unpack('extents', None)
			if extents is None:
				'assign to 0...?'
				return None
			
			# TODO resize the local extents to the given extents (fit)
			#raise NotImplementedError()


		@ENV.SnapChannel
		def set(self, MSG):
			"""(
			width=int|float?, height=int|float?, depth=int|float,
			size=tuple|list(2 * int|float)?, dimensions=tuple|list(3 * int|float)?,
			x=int|float?, y=int|float?, z=int|float?,
			position=list|tuple(2 * int|float)?, origin=list|tuple(3 * int|float)?,
			extents=snap_extents_t?,
			)"""
			return SnapMatrix.set(self, MSG)

			# TODO use with to turn off changed emissions?  make that standard on SnapNode?

			"""
			ext = self['extents'] or snap_extents_t()
			changed = False

			for attr,value in MSG.kwargs.items():

				if attr in ('width', 'height', 'depth'):
					axis = ['width','height','depth'].index(attr)
					if not value:
						value = 0
					ext[axis+3] = ext[axis] + value
					changed = True

				elif attr in ('size', 'dimensions'):
					axes = [0,1,2] if attr == 'dimensions' else [0,1]
					if not value:
						value = [0,0,0]
					for axis in axes:
						ext[axis+3] = ext[axis] + value[axis]
					changed = True

				elif attr in ('x','y','z'):
					axis = ['x','y','z'].index(attr)
					if not value:
						value = 0
					# this behaves like a translation, moving the whole rect
					diff = ext[axis+3]-ext[axis]
					ext[axis] = value
					ext[axis+3] = value + diff
					changed = True

				elif attr in ('position', 'origin'):
					axes = [0,1,2] if attr == 'origin' else [0,1]
					if not value:
						value = [0,0,0]
					for axis in axes:
						diff = ext[axis+3] - ext[axis]
						ext[axis] = value[axis]
						ext[axis+3] = value[axis] + diff
					changed=True

				elif attr in ('extents', '_extents_'):
					ext[:] = value
					changed = True

				else:
					ENV.snap_warning("unsupported setting", attr)
					#SnapMatrix.set(self, **{attr:value})

			if changed:
				data['extents'] = ext
				if ext[3] <= ext[0]: ext[3] = ext[0]+1
				if ext[4] <= ext[1]: ext[4] = ext[1]+1
				if ext[5] <= ext[2]: ext[5] = ext[2]+1
				#snap_send(self.changed, extents=ext) XXX TODO
				#snap_emit(self, "CHANGED", extents=ext)
				self.changed(extents=ext)
			"""

		def __repr__(self):
			return SnapMatrix.__repr__(self) # TODO

		def __init__(self, **SETTINGS):
			SnapMatrix.__init__(self, **SETTINGS)

			#if extents is not None:
			#	self['extents'] = snap_extents_t(*extents)
			#else:
			#	self['extents'] = None # assigned means it is known for this object; cached; ie don't need to do a full walk to determine the extents...  #snap_extents_t(0,0,0, 0,0,0)



	ENV.SnapMetrics = SnapMetrics


def main(ENV):

	ENV.SnapMetrics()

	ENV.snap_out('ok')

